src/components/implementation/torrent/ramfs/ramfs.c

```diff
@@ -186,6 +278,160 @@ twrite(spdid_t spdid, td_t td, int cbid, int sz)
 	return ret;
 }
 
+int 
+twritep(spdid_t spdid, td_t td, int cb, int start, int sz)
+{
+	int ret = -1, left;
+	struct torrent *t;
+	struct fsobj *fso;
+
+	if (!sz) {return -EINVAL;}
+
+	if (tor_isnull(td)) {return -EINVAL;}
+
+	LOCK();
+
+	t = tor_lookup(td);
+	if (!t) ERR_THROW(-EINVAL, done);
+	assert(t->data);
+	if (!(t->flags & TOR_WRITE)) ERR_THROW(-EACCES, done);
+
+	fso = t->data;
+	// we aren't tracking these right now. Need to either fix or write equivalent checks.
+	// fso->size may be irrelevant now. Or not?
+	//assert(fso->size <= fso->allocated);
+	//assert(t->offset <= fso->size);
+
+	unsigned int total_offset = 0;
+	struct file_data *current = (struct file_data*) fso->data;
+	
+	/* file does not exist yet */
+	if (current == NULL) {
+		struct file_data *new_node;
+		file_data_alloc(&new_node, cb, 0, sz, start, sz, NULL);
+
+		fso->data = new_node;
+
+		t->offset += sz;
+		fso->size = t->offset;
+
+		ret = sz;
+
+		goto done;
+	}
+	
+	unsigned int curr_off = 0;
+	struct file_data *prev = NULL;
+
+	/* iterate through and try to find where to write to */
+	while (current != NULL &&
+		(total_offset + current->len) <= t->offset) {
+		total_offset += current->len;
+
+		prev = current;
+		current = current->next;
+	}
+
+	curr_off = t->offset - total_offset;
+
+	if (current == NULL) {
```

Comment:
did you ever change all of this logic to simplify it using the technique we talked about???